vm/compiler/src/coinbase_puzzle/mod.rs

```diff
@@ -0,0 +1,363 @@
+        }
+
+        // Return the accumulated proof.
+        Ok(CoinbaseSolution::new(partial_solutions, proof))
+    }
+
+    /// Returns `true` if the coinbase solution is valid.
+    pub fn verify(
+        &self,
+        coinbase_solution: &CoinbaseSolution<N>,
+        epoch_challenge: &EpochChallenge<N>,
+        coinbase_target: u64,
+        proof_target: u64,
+    ) -> Result<bool> {
+        // Ensure the coinbase solution is not empty.
+        if coinbase_solution.is_empty() {
+            bail!("The coinbase solution does not contain any partial solutions");
+        }
+
+        // Ensure the number of partial solutions does not exceed `MAX_PROVER_SOLUTIONS`.
+        if coinbase_solution.len() > MAX_PROVER_SOLUTIONS {
+            bail!(
+                "The coinbase solution exceeds the allowed number of partial solutions. ({} > {MAX_PROVER_SOLUTIONS})",
+                coinbase_solution.len()
+            );
+        }
+
+        // Ensure the coinbase proof is non-hiding.
+        if coinbase_solution.proof().is_hiding() {
+            bail!("The coinbase proof must be non-hiding");
+        }
+
+        // Ensure the coinbase proof meets the required coinbase target.
+        if coinbase_solution.to_cumulative_target()? < coinbase_target as u128 {
+            bail!("The coinbase proof does not meet the coinbase target");
+        }
+
+        // Compute the prover polynomials.
+        let prover_polynomials = cfg_iter!(coinbase_solution.partial_solutions())
+            // Ensure that each of the prover solutions meets the required proof target.
+            .map(|solution| match solution.to_target()? >= proof_target {
+                // Compute the prover polynomial.
+                true => solution.to_prover_polynomial(epoch_challenge),
+                false => bail!("Prover puzzle does not meet the proof target requirements."),
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        // Compute the challenge points.
+        let mut challenge_points =
+            hash_commitments(coinbase_solution.partial_solutions().iter().map(|solution| *solution.commitment()))?;
+        ensure!(
+            challenge_points.len() == coinbase_solution.partial_solutions().len() + 1,
+            "Invalid number of challenge points"
+        );
+
+        // Pop the last challenge point as the accumulator challenge point.
+        let accumulator_point = match challenge_points.pop() {
+            Some(point) => point,
+            None => bail!("Missing the accumulator challenge point"),
+        };
+
+        // Compute the accumulator evaluation.
+        let mut accumulator_evaluation = cfg_iter!(prover_polynomials)
+            .zip_eq(&challenge_points)
+            .fold(<N::PairingCurve as PairingEngine>::Fr::zero, |accumulator, (prover_polynomial, challenge_point)| {
+                accumulator + (prover_polynomial.evaluate(accumulator_point) * challenge_point)
+            })
+            .sum();
```

Comment:
same question as in https://github.com/AleoHQ/snarkVM/pull/1100/files#r996828432